### PR TITLE
feat: return images from resources when sync occurs

### PR DESCRIPTION
**Pull request by Aaron-9900**
Implementing the Gitops side of this issue: https://github.com/argoproj/argo-cd/issues/20896

Adds an api to `kube.go` to get the images from a resource. Adds those images to `ResourceSyncResult`. This API accounts for how different k8s resources define the images.

### CODE REVIEW

File: pkg/utils/kube/kube.go [hunk 1]
```diff
@@ -404,6 +404,34 @@ func GetDeploymentReplicas(u *unstructured.Unstructured) *int64 {
 	return &val
 }
 
+func GetResourceImages(u *unstructured.Unstructured) []string {
```
Comment (andrea-mosk) [inline pkg/utils/kube/kube.go:407]: For [CronJobs](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#example) it seems that the image is under `spec.jobTemplate.spec.template.spec.containers`. Should we also check there?

Comment (Aaron-9900) [reply pkg/utils/kube/kube.go:407]: Done in 6b9063a72199550b4fd51beabdabdc3104b1f7fb

File: pkg/utils/kube/kube_test.go [hunk 2]
```diff
@@ -167,6 +166,170 @@ spec:
...
 
+func TestGetResourceImages(t *testing.T) {
+	type testCase struct {
+		manifest    []byte
+		expected    []string
+		description string
+	}
+
+	testCases := []testCase{
+		{
```
Comment (andrea-mosk) [inline pkg/utils/kube/kube_test.go:177]:
```suggestion
	type testCase struct {
		manifest    []byte
		expected    []string
		description string
	}{
		{
```

We can probably add the no images tests here right? By just expecting an empty/nil array?

**Review by andrea-mosk — APPROVED** [review]

**Review by rumstead — APPROVED** [review]
LGTM

**Comment by rumstead** [comment]
Can you create a PR to Argo CD updating the `go.mod` to your branch? That will allow us to validate the e2e tests run and pass.

**Comment by Aaron-9900** [comment]
@rumstead done in https://github.com/argoproj/argo-cd/pull/20954, tests are now passing

**Comment by Aaron-9900** [comment]
Added new benchmark test for the sync operation. These are the results:

command: `go test -bench=. -benchtime=2s -count 10`

<details>
<summary>3 images</summary>

```

—— new

goos: darwin
goarch: arm64
pkg: github.com/argoproj/gitops-engine/pkg/sync
cpu: Apple M1 Max
BenchmarkSync-10    	   63940	     37651 ns/op
BenchmarkSync-10    	   60710	     40457 ns/op
BenchmarkSync-10    	   56287	     40774 ns/op
BenchmarkSync-10    	   60465	     40192 ns/op
BenchmarkSync-10    	   60189	     40748 ns/op
BenchmarkSync-10    	   59324	     38114 ns/op
BenchmarkSync-10    	   61059	     37721 ns/op
BenchmarkSync-10    	   64406	     38233 ns/op
BenchmarkSync-10    	   58512	     37904 ns/op
BenchmarkSync-10    	   60928	     38100 ns/op

—— old
goos: darwin
goarch: arm64
pkg: github.com/argoproj/gitops-engine/pkg/sync
cpu: Apple M1 Max
BenchmarkSync-10    	   60399	     39880 ns/op
BenchmarkSync-10    	   61432	     35195 ns/op
BenchmarkSync-10    	   72246	     34367 ns/op
BenchmarkSync-10    	   70420	     34345 ns/op
BenchmarkSync-10    	   72147	     33829 ns/op
BenchmarkSync-10    	   70708	     37917 ns/op
BenchmarkSync-10    	   59630	     34623 ns/op
BenchmarkSync-10    	   71232	     33385 ns/op
BenchmarkSync-10    	   72580	     33692 ns/op
BenchmarkSync-10    	   69116	     34070 ns/op

—— comparison
goos: darwin
goarch: arm64
pkg: github.com/argoproj/gitops-engine/pkg/sync
cpu: Apple M1 Max
        │   old.txt    │               new.txt               │
        │    sec/op    │   sec/op     vs base                │
Sync-10   34.36µ ± 10%   38.17µ ± 7%  +11.11% (p=0.001 n=10)	
```

</details>

<details>
<summary>5 images </summary>

```
—— new
goos: darwin
goarch: arm64
pkg: github.com/argoproj/gitops-engine/pkg/sync
cpu: Apple M1 Max
BenchmarkSync-10    	   59912	     42840 ns/op
BenchmarkSync-10    	   50362	     44076 ns/op
BenchmarkSync-10    	   57453	     44372 ns/op
BenchmarkSync-10    	   57046	     41663 ns/op
BenchmarkSync-10    	   56764	     44259 ns/op
BenchmarkSync-10    	   58297	     42187 ns/op
BenchmarkSync-10    	   55392	     43204 ns/op
BenchmarkSync-10    	   58656	     40735 ns/op
BenchmarkSync-10    	   62046	     42186 ns/op
BenchmarkSync-10    	   56892	     45048 ns/op
—— old
goos: darwin
goarch: arm64
pkg: github.com/argoproj/gitops-engine/pkg/sync
cpu: Apple M1 Max
BenchmarkSync-10    	   72872	     33205 ns/op
BenchmarkSync-10    	   73526	     33451 ns/op
BenchmarkSync-10    	   68599	     33591 ns/op
BenchmarkSync-10    	   70236	     33961 ns/op
BenchmarkSync-10    	   60986	     35286 ns/op
BenchmarkSync-10    	   71302	     34127 ns/op
BenchmarkSync-10    	   70725	     34121 ns/op
BenchmarkSync-10    	   72074	     34371 ns/op
BenchmarkSync-10    	   72940	     37116 ns/op
BenchmarkSync-10    	   68548	     35546 ns/op
— Comparison
goos: darwin
goarch: arm64
pkg: github.com/argoproj/gitops-engine/pkg/sync
cpu: Apple M1 Max
        │   old.txt   │               new.txt               │
        │   sec/op    │   sec/op     vs base                │
Sync-10   34.12µ ± 4%   43.02µ ± 3%  +26.08% (p=0.000 n=10)
```
</details>

<details>
<summary>10 images</summary>

```
— new
goos: darwin
goarch: arm64
pkg: github.com/argoproj/gitops-engine/pkg/sync
cpu: Apple M1 Max
BenchmarkSync-10    	   52508	     45693 ns/op
BenchmarkSync-10    	   55818	     43287 ns/op
BenchmarkSync-10    	   55305	     46177 ns/op
BenchmarkSync-10    	   54345	     44417 ns/op
BenchmarkSync-10    	   54358	     46124 ns/op
BenchmarkSync-10    	   51956	     44968 ns/op
BenchmarkSync-10    	   49981	     48196 ns/op
BenchmarkSync-10    	   48973	     49551 ns/op
BenchmarkSync-10    	   49080	     49411 ns/op
BenchmarkSync-10    	   52101	     46850 ns/op

—— old
goos: darwin
goarch: arm64
pkg: github.com/argoproj/gitops-engine/pkg/sync
cpu: Apple M1 Max
BenchmarkSync-10    	   70117	     36253 ns/op
BenchmarkSync-10    	   66276	     35481 ns/op
BenchmarkSync-10    	   73176	     33779 ns/op
BenchmarkSync-10    	   71619	     34551 ns/op
BenchmarkSync-10    	   74025	     33803 ns/op
BenchmarkSync-10    	   73438	     33614 ns/op
BenchmarkSync-10    	   72687	     33609 ns/op
BenchmarkSync-10    	   70243	     33591 ns/op
BenchmarkSync-10    	   70620	     34983 ns/op
BenchmarkSync-10    	   70617	     33780 ns/op
PASS
ok  	github.com/argoproj/gitops-engine/pkg/sync	119.814s

—— comparison

goos: darwin
goarch: arm64
pkg: github.com/argoproj/gitops-engine/pkg/sync
cpu: Apple M1 Max
        │   old.txt   │               new.txt               │
        │   sec/op    │   sec/op     vs base                │
Sync-10   33.79µ ± 5%   46.15µ ± 7%  +36.57% (p=0.000 n=10)
```
</details>

Overall, we see a small increase in time/operation for a small amount of images (11% for 3 images), and it gets larger as the images grow (36% for 10 images). This is expected, since we are now iterating over images. Even with this growth, the total value is still quite low.

File: pkg/sync/common/types.go [hunk 3]
```diff
@@ -136,6 +136,8 @@ func NewHookDeletePolicy(p string) (HookDeletePolicy, bool) {
 type ResourceSyncResult struct {
 	// holds associated resource key
 	ResourceKey kube.ResourceKey
+	// holds the images associated with the resource
+	Images []string
```
Comment (crenshaw-dev) [inline pkg/sync/common/types.go:140]:
```suggestion
	// Images holds the images associated with the resource. These images are collected on a best-effort basis
	// from fields used by known workload resources. This does not necessarily reflect the exact list of images
	// used by workloads in the application.
	Images []string
```

**Review by crenshaw-dev — APPROVED** [review]
